File: miner-engine/r3mes/debug_config.py
"""
Debug Configuration System for R3MES Miner Engine

Provides centralized debug configuration management with environment variable support.
"""

import os
from typing import Dict, Optional, Set
from dataclasses import dataclass, field
from enum import Enum


class DebugLevel(str, Enum):
    """Debug verbosity level"""
    MINIMAL = "minimal"  # Only critical errors and performance metrics
    STANDARD = "standard"  # Detailed logging, state inspection, performance profiling
    VERBOSE = "verbose"  # All system internals, trace logs, internal state dumps


@dataclass
class DebugConfig:
    """Debug configuration for the miner engine component"""
    
    # Global debug mode enabled
    enabled: bool = False
    
    # Debug verbosity level
    level: DebugLevel = DebugLevel.STANDARD
    
    # Enabled components (comma-separated: blockchain,backend,miner,launcher,frontend)
    # Empty or "*" means all components
    components: Set[str] = field(default_factory=set)
    
    # Feature flags
    logging: bool = True
    profiling: bool = True
    state_inspection: bool = True
    trace: bool = True
    
    # Logging configuration
    log_level: str = "TRACE"  # TRACE, DEBUG, INFO, WARN, ERROR
    log_format: str = "json"  # json, text
    log_file: Optional[str] = None
    
    # Performance profiling configuration
    profile_output: str = "~/.r3mes/profiles"
    profile_interval: int = 60  # seconds
    
    # Trace configuration
    trace_enabled: bool = True
    trace_buffer_size: int = 10000
    trace_export_path: str = "~/.r3mes/traces"
    
    # Miner-specific debug flags
    training_loop_debug: bool = True  # Debug training loop iterations
    gradient_debug: bool = True  # Debug gradient computations
    blockchain_interaction_debug: bool = True  # Debug blockchain interactions
    ipfs_debug: bool = True  # Debug IPFS operations
    
    def is_component_enabled(self, component: str) -> bool:
        """Check if debug is enabled for a specific component"""
        if not self.enabled:
            return False
        # If no components specified or "*" is present, enable all
        if not self.components or "*" in self.components:
            return True
        return component.lower() in self.components
    
    def is_miner_enabled(self) -> bool:
        """Check if debug is enabled for miner component"""
        return self.is_component_enabled("miner")


def _expand_path(path: str) -> str:
    """Expand user home directory in path"""
    if path.startswith("~/"):
        home_dir = os.path.expanduser("~")
        return path.replace("~/", f"{home_dir}/", 1)
    return path


def _get_env_bool(key: str, default: bool = False) -> bool:
    """Read a boolean environment variable"""
    value = os.getenv(key, "")
    if not value:
        return default
    return value.lower() == "true"


def _get_env_int(key: str, default: int) -> int:
    """Read an integer environment variable"""
    value = os.getenv(key, "")
    if not value:
        return default
    try:
        return int(value)
    except ValueError:
        return default


def load_debug_config() -> DebugConfig:
    """
    Load debug configuration from environment variables.
    
    Returns:
        DebugConfig instance with loaded configuration
    """
    config = DebugConfig()
    
    # Check if debug mode is enabled
    debug_mode = os.getenv("R3MES_DEBUG_MODE", "").lower()
    config.enabled = debug_mode == "true"
    
    if not config.enabled:
        # Return default (disabled) config
        return config
    
    # Load debug level
    level_str = os.getenv("R3MES_DEBUG_LEVEL", "").lower()
    try:
        config.level = DebugLevel(level_str) if level_str else DebugLevel.VERBOSE
    except ValueError:
        # Default to verbose if level is not valid
        config.level = DebugLevel.VERBOSE
    
    # Load enabled components
    components_str = os.getenv("R3MES_DEBUG_COMPONENTS", "")
    if not components_str:
        # Default to all components if not specified
        config.components = {"*"}
    else:
        components = [c.strip().lower() for c in components_str.split(",") if c.strip()]
        config.components = set(components)
    
    # Load feature flags (default to true if debug mode is enabled)
    config.logging = _get_env_bool("R3MES_DEBUG_LOGGING", True)
    config.profiling = _get_env_bool("R3MES_DEBUG_PROFILING", True)
    config.state_inspection = _get_env_bool("R3MES_DEBUG_STATE_INSPECTION", True)
    config.trace = _get_env_bool("R3MES_DEBUG_TRACE", True)
    
    # Load logging configuration
    log_level = os.getenv("R3MES_DEBUG_LOG_LEVEL", "TRACE").upper()
    valid_log_levels = {"TRACE", "DEBUG", "INFO", "WARN", "ERROR"}
    if log_level in valid_log_levels:
        config.log_level = log_level
    else:
        config.log_level = "TRACE"  # Default
    
    log_format = os.getenv("R3MES_DEBUG_LOG_FORMAT", "json").lower()
    if log_format in ("json", "text"):
        config.log_format = log_format
    else:
        config.log_format = "json"  # Default
    
    log_file = os.getenv("R3MES_DEBUG_LOG_FILE", "")
    if log_file:
        config.log_file = _expand_path(log_file)
    else:
        config.log_file = _expand_path("~/.r3mes/debug.log")
    
    # Load performance profiling configuration
    profile_output = os.getenv("R3MES_DEBUG_PROFILE_OUTPUT", "")
    if profile_output:
        config.profile_output = _expand_path(profile_output)
    else:
        config.profile_output = _expand_path("~/.r3mes/profiles")
    
    config.profile_interval = _get_env_int("R3MES_DEBUG_PROFILE_INTERVAL", 60)
    
    # Load trace configuration
    config.trace_enabled = _get_env_bool("R3MES_DEBUG_TRACE_ENABLED", True)
    config.trace_buffer_size = _get_env_int("R3MES_DEBUG_TRACE_BUFFER_SIZE", 10000)
    
    trace_export_path = os.getenv("R3MES_DEBUG_TRACE_EXPORT_PATH", "")
    if trace_export_path:
        config.trace_export_path = _expand_path(trace_export_path)
    else:
        config.trace_export_path = _expand_path("~/.r3mes/traces")
    
    # Load miner-specific debug flags
    config.training_loop_debug = _get_env_bool("R3MES_DEBUG_TRAINING_LOOP", True)
    config.gradient_debug = _get_env_bool("R3MES_DEBUG_GRADIENT", True)
    config.blockchain_interaction_debug = _get_env_bool("R3MES_DEBUG_BLOCKCHAIN", True)
    config.ipfs_debug = _get_env_bool("R3MES_DEBUG_IPFS", True)
    
    return config


def validate_debug_config(is_production: bool = False) -> Optional[str]:
    """
    Validate debug configuration for security.
    
    Args:
        is_production: Whether running in production environment
        
    Returns:
        Error message if validation fails, None otherwise
    """
    if not is_production:
        # In development/testing, debug mode is allowed
        return None
    
    # In production, check if debug mode is enabled
    debug_mode = os.getenv("R3MES_DEBUG_MODE", "").lower()
    if debug_mode == "true":
        return (
            "SECURITY ERROR: R3MES_DEBUG_MODE=true is set in production environment. "
            "Debug mode should only be used in development/testing. "
            "Please unset R3MES_DEBUG_MODE environment variable before running in production. "
            "If you need production debugging, use R3MES_DEBUG_MODE=minimal with explicit component flags."
        )
    
    # Check for verbose debug level in production (even if explicitly enabled)
    debug_level = os.getenv("R3MES_DEBUG_LEVEL", "").lower()
    if debug_level == "verbose":
        return (
            "SECURITY WARNING: R3MES_DEBUG_LEVEL=verbose is not recommended in production. "
            "Consider using 'standard' or 'minimal' level instead."
        )
    
    return None


# Global debug config instance (lazy-loaded)
_global_debug_config: Optional[DebugConfig] = None


def get_debug_config() -> DebugConfig:
    """
    Get the global debug configuration (cached).
    
    Returns:
        DebugConfig instance
    """
    global _global_debug_config
    if _global_debug_config is None:
        _global_debug_config = load_debug_config()
    return _global_debug_config

File: remes/x/remes/keeper/debug_config.go
package keeper

import (
	"fmt"
	"os"
	"strings"
)

// DebugLevel represents the verbosity level for debug mode
type DebugLevel string

const (
	DebugLevelMinimal  DebugLevel = "minimal"  // Only critical errors and performance metrics
	DebugLevelStandard DebugLevel = "standard" // Detailed logging, state inspection, performance profiling
	DebugLevelVerbose  DebugLevel = "verbose"  // All system internals, trace logs, internal state dumps
)

// DebugConfig holds debug configuration for the keeper
type DebugConfig struct {
	// Enabled indicates if debug mode is globally enabled
	Enabled bool

	// Level specifies the verbosity level
	Level DebugLevel

	// Components specifies which components should have debug enabled (comma-separated)
	// Valid values: blockchain,backend,miner,launcher,frontend
	// Empty or "*" means all components
	Components map[string]bool

	// Features
	Logging          bool // Enable enhanced logging
	Profiling        bool // Enable performance profiling
	StateInspection  bool // Enable state inspection
	Trace            bool // Enable distributed tracing

	// Logging configuration
	LogLevel string // TRACE, DEBUG, INFO, WARN, ERROR
	LogFormat string // json, text
	LogFile   string // Path to log file (optional)

	// Performance profiling configuration
	ProfileOutput   string // Directory for profile outputs
	ProfileInterval int    // Interval in seconds for periodic profiling

	// Trace configuration
	TraceEnabled   bool   // Enable trace collection
	TraceBufferSize int   // Maximum number of traces to keep in memory
	TraceExportPath string // Path to export traces
}

// IsComponentEnabled checks if debug is enabled for a specific component
func (dc *DebugConfig) IsComponentEnabled(component string) bool {
	if !dc.Enabled {
		return false
	}
	// If no components specified or "*" is present, enable all
	if len(dc.Components) == 0 || dc.Components["*"] {
		return true
	}
	return dc.Components[strings.ToLower(component)]
}

// IsBlockchainEnabled checks if debug is enabled for blockchain component
func (dc *DebugConfig) IsBlockchainEnabled() bool {
	return dc.IsComponentEnabled("blockchain")
}

// LoadDebugConfig loads debug configuration from environment variables
func LoadDebugConfig() (*DebugConfig, error) {
	config := &DebugConfig{
		Components: make(map[string]bool),
	}

	// Check if debug mode is enabled
	debugMode := os.Getenv("R3MES_DEBUG_MODE")
	config.Enabled = strings.ToLower(debugMode) == "true"

	if !config.Enabled {
		// Return default (disabled) config
		return config, nil
	}

	// Load debug level
	levelStr := strings.ToLower(os.Getenv("R3MES_DEBUG_LEVEL"))
	switch levelStr {
	case "minimal":
		config.Level = DebugLevelMinimal
	case "standard":
		config.Level = DebugLevelStandard
	case "verbose":
		config.Level = DebugLevelVerbose
	default:
		// Default to verbose if level is not specified
		config.Level = DebugLevelVerbose
	}

	// Load enabled components
	componentsStr := os.Getenv("R3MES_DEBUG_COMPONENTS")
	if componentsStr == "" {
		// Default to all components if not specified
		config.Components["*"] = true
	} else {
		components := strings.Split(componentsStr, ",")
		for _, comp := range components {
			comp = strings.TrimSpace(strings.ToLower(comp))
			if comp != "" {
				config.Components[comp] = true
			}
		}
	}

	// Load feature flags (default to true if debug mode is enabled)
	config.Logging = getEnvBool("R3MES_DEBUG_LOGGING", true)
	config.Profiling = getEnvBool("R3MES_DEBUG_PROFILING", true)
	config.StateInspection = getEnvBool("R3MES_DEBUG_STATE_INSPECTION", true)
	config.Trace = getEnvBool("R3MES_DEBUG_TRACE", true)

	// Load logging configuration
	config.LogLevel = strings.ToUpper(os.Getenv("R3MES_DEBUG_LOG_LEVEL"))
	if config.LogLevel == "" {
		config.LogLevel = "TRACE" // Default to TRACE in debug mode
	}
	// Validate log level
	validLogLevels := map[string]bool{
		"TRACE": true,
		"DEBUG": true,
		"INFO":  true,
		"WARN":  true,
		"ERROR": true,
	}
	if !validLogLevels[config.LogLevel] {
		return nil, fmt.Errorf("invalid R3MES_DEBUG_LOG_LEVEL: %s (valid: TRACE, DEBUG, INFO, WARN, ERROR)", config.LogLevel)
	}

	config.LogFormat = strings.ToLower(os.Getenv("R3MES_DEBUG_LOG_FORMAT"))
	if config.LogFormat == "" {
		config.LogFormat = "json" // Default to JSON for structured logging
	}
	if config.LogFormat != "json" && config.LogFormat != "text" {
		return nil, fmt.Errorf("invalid R3MES_DEBUG_LOG_FORMAT: %s (valid: json, text)", config.LogFormat)
	}

	config.LogFile = os.Getenv("R3MES_DEBUG_LOG_FILE")
	// Default log file location if not specified
	if config.LogFile == "" {
		homeDir, err := os.UserHomeDir()
		if err == nil {
			config.LogFile = fmt.Sprintf("%s/.r3mes/debug.log", homeDir)
		}
	}

	// Load performance profiling configuration
	config.ProfileOutput = os.Getenv("R3MES_DEBUG_PROFILE_OUTPUT")
	if config.ProfileOutput == "" {
		homeDir, err := os.UserHomeDir()
		if err == nil {
			config.ProfileOutput = fmt.Sprintf("%s/.r3mes/profiles", homeDir)
		}
	}
	config.ProfileInterval = getEnvInt("R3MES_DEBUG_PROFILE_INTERVAL", 60)

	// Load trace configuration
	config.TraceEnabled = getEnvBool("R3MES_DEBUG_TRACE_ENABLED", true)
	config.TraceBufferSize = getEnvInt("R3MES_DEBUG_TRACE_BUFFER_SIZE", 10000)
	config.TraceExportPath = os.Getenv("R3MES_DEBUG_TRACE_EXPORT_PATH")
	if config.TraceExportPath == "" {
		homeDir, err := os.UserHomeDir()
		if err == nil {
			config.TraceExportPath = fmt.Sprintf("%s/.r3mes/traces", homeDir)
		}
	}

	return config, nil
}

// ValidateDebugConfig validates debug configuration for security
// This should be called to ensure debug mode is not accidentally enabled in production
func ValidateDebugConfig(isProduction bool) error {
	if !isProduction {
		// In development/testing, debug mode is allowed
		return nil
	}

	// In production, check if debug mode is enabled
	debugMode := os.Getenv("R3MES_DEBUG_MODE")
	if strings.ToLower(debugMode) == "true" {
		return fmt.Errorf(
			"SECURITY ERROR: R3MES_DEBUG_MODE=true is set in production environment. " +
				"Debug mode should only be used in development/testing. " +
				"Please unset R3MES_DEBUG_MODE environment variable before running in production. " +
				"If you need production debugging, use R3MES_DEBUG_MODE=minimal with explicit component flags",
		)
	}

	// Check for verbose debug level in production (even if explicitly enabled)
	debugLevel := strings.ToLower(os.Getenv("R3MES_DEBUG_LEVEL"))
	if debugLevel == "verbose" {
		return fmt.Errorf(
			"SECURITY WARNING: R3MES_DEBUG_LEVEL=verbose is not recommended in production. " +
				"Consider using 'standard' or 'minimal' level instead",
		)
	}

	return nil
}

// getEnvBool reads a boolean environment variable
// Returns defaultValue if the variable is not set or cannot be parsed
func getEnvBool(key string, defaultValue bool) bool {
	value := os.Getenv(key)
	if value == "" {
		return defaultValue
	}
	return strings.ToLower(value) == "true"
}

// getEnvInt reads an integer environment variable
// Returns defaultValue if the variable is not set or cannot be parsed
func getEnvInt(key string, defaultValue int) int {
	value := os.Getenv(key)
	if value == "" {
		return defaultValue
	}
	var result int
	_, err := fmt.Sscanf(value, "%d", &result)
	if err != nil {
		return defaultValue
	}
	return result
}

// GetDebugConfig returns the global debug configuration
// This is a convenience function that caches the config (thread-safe for read-only access)
var globalDebugConfig *DebugConfig
var debugConfigInitialized bool

func GetDebugConfig() *DebugConfig {
	if !debugConfigInitialized {
		// Load config on first access (not thread-safe, but acceptable for initialization)
		config, err := LoadDebugConfig()
		if err != nil {
			// On error, return disabled config
			globalDebugConfig = &DebugConfig{
				Enabled:    false,
				Level:      DebugLevelStandard,
				Components: make(map[string]bool),
			}
		} else {
			globalDebugConfig = config
		}
		debugConfigInitialized = true
	}
	return globalDebugConfig
}
